sdk/vppagent/acl.go

```diff
@@ -0,0 +1,319 @@
+		err = fmt.Errorf("found empty source name")
+		logrus.Errorf("Invalid connection data: %v", err)
+		return nil, err
+	}
+
+	connectionData.DataChange, err = a.appendDataChange(connectionData.DataChange, connectionData.SrcName)
+	if err != nil {
+		logrus.Error(err)
+		return nil, err
+	}
+
+	a.Connections[incomingConnection.GetId()] = connectionData
+
+	return incomingConnection, nil
+}
+
+// Close implements the close handler
+func (a *ACL) Close(ctx context.Context, connection *connection.Connection) (*empty.Empty, error) {
+	if a.GetNext() != nil {
+		return a.GetNext().Close(ctx, connection)
+	}
+	return &empty.Empty{}, nil
+}
+
+// GetOpaque will return the corresponding connection data
+func (a *ACL) GetOpaque(incoming interface{}) interface{} {
+	incomingConnection := incoming.(*connection.Connection)
+	if connectionData, ok := a.Connections[incomingConnection.GetId()]; ok {
+		return connectionData
+	}
+	logrus.Errorf("GetOpaque outgoing not found for %v", incomingConnection)
+	return nil
+}
+
+// Name returns the composite name
+func (a *ACL) Name() string {
+	return "acl"
+}
+
+// NewACL creates an ACL
+func NewACL(configuration *common.NSConfiguration, rules map[string]string) *ACL {
+	// ensure the env variables are processed
+	if configuration == nil {
+		configuration = &common.NSConfiguration{}
+	}
+	configuration.CompleteNSConfiguration()
+
+	return &ACL{
+		Rules:       rules,
+		Connections: map[string]*ConnectionData{},
+	}
+}
+
+func (a *ACL) appendDataChange(rv *configurator.Config, ingressInterface string) (*configurator.Config, error) {
+	if rv == nil {
+		rv = &configurator.Config{}
+	}
+	if rv.VppConfig == nil {
+		rv.VppConfig = &vpp.ConfigData{}
+	}
+
+	rules := []*acl.ACL_Rule{}
+
+	for _, rule := range a.Rules {
+		parsed := tools.ParseKVStringToMap(rule, ",", "=")
+
+		action, err := getAction(parsed)
+		if err != nil {
+			logrus.Errorf("Parsing rule %s failed with %v", rule, err)
+			return nil, err
+		}
+
+		match, err := getMatch(parsed)
+		match.Action = action
+		if err != nil {
+			logrus.Errorf("Parsing rule %s failed with %v", rule, err)
+			return nil, err
+		}
+
+		rules = append(rules, match)
+
+		rv.VppConfig.Acls = append(rv.VppConfig.Acls, &acl.ACL{
+			Name:  "IngressACL",
+			Rules: rules,
+			Interfaces: &acl.ACL_Interfaces{
+				Egress: []string{},
+				Ingress: []string{
+					ingressInterface,
+				},
+			},
+		})
+	}
+
+	return rv, nil
+}
+
+func getAction(parsed map[string]string) (acl.ACL_Rule_Action, error) {
+	actionName, ok := parsed["action"]
+	if !ok {
+		return acl.ACL_Rule_Action(0), fmt.Errorf("rule should have 'action' set")
+	}
+	action, ok := acl.ACL_Rule_Action_value[strings.ToUpper(actionName)]
+	if !ok {
+		return acl.ACL_Rule_Action(0), fmt.Errorf("rule should have a valid 'action'")
+	}
+	return acl.ACL_Rule_Action(action), nil
+}
+
+func getIP(parsed map[string]string) (*acl.ACL_Rule_IpRule_Ip, error) {
+	dstNet, dstNetOk := parsed["dstnet"]
```

Comment:
Should we put "dstnet", "stcnet" into constants?

Reply:
Done